keps/sig-storage/3756-volume-reconstruction/README.md

```diff
@@ -0,0 +1,1030 @@
+      and reconstructs **only** volumes that are not already in  ASW.
+      In addition, volumes that are in DSW are reconstructed, but not added to
+      ASW (If a volume is in DSW, we expect that it reaches ASW during step 3.)
+      * `devicePath` of reconstructed volumes is populated from
+        `node.status.attachedVolumes` right away.
+      * In the next reconciliation loop, reconstructed volumes that are not in
+        DSW are finally unmounted in step 1. above.
+      * There is a workaround to add a reconstructed volume to ASW when it was
+        initially in DSW, but all pods that used the volume were deleted before
+        the volume was mounted and reached ASW.
+        ([#110670](https://github.com/kubernetes/kubernetes/pull/110670))
+   2. VolumeManager reports all reconstructed volumes in
+      `node.status.volumesInUse` (that's why VolumeManager reconstructs volumes,
+      even if it does not add them to DSW).
+   3. For volumes that failed reconstruction kubelet cannot call appropriate
+      volume plugin to unmount them. Kubelet at least tries to unmount the
+      directory and clean up any orphan files there.
+
+#### Observability
+
+Today, any errors during volume reconstruction are exposed only as log messages.
+We propose adding these new metrics, both to the old and new VolumeManager code:
+
+* `reconstructed_volumes_total` with label `result={success, error}`: nr. of
+  successfully / unsuccessfully reconstructed volumes.
+  * In the new VolumeManager code, this will include all volume mounts in
+    `/var/lib/kubelet/pods/*/volumes`
+  * In the old VolumeManager it will include only volumes that were not already
+    in ASW (those are not reconstructed).
+* `force_cleaned_failed_volumes_total` with label `result={success, error}`: nr.
+  of successful / unsuccessful cleanups of volumes that failed reconstruction.
+* `orphaned_volumes_cleanup_errors_total`: nr. of reports
+  like `orphaned pod "<uid>" found, but XYZ failed`
+  ([example](https://github.com/kubernetes/kubernetes/blob/4fac7486d41c033d6bba9dfeda2356e8189035cd/pkg/kubelet/kubelet_volumes.go#L215)).
+  These messages can be a symptom of failed reconstruction (e.g.
+  [#105536](https://github.com/kubernetes/kubernetes/issues/105536)).
+  Note that kubelet logs this periodically and bumping this metric periodically
+  would not be useful.
+  [`cleanupOrphanedPodDirs`](https://github.com/kubernetes/kubernetes/blob/4fac7486d41c033d6bba9dfeda2356e8189035cd/pkg/kubelet/kubelet_volumes.go#L168)
+  needs to be changed to collect errors found during
+  one `/var/lib/kubelet/pods/` check and report collected "nr of errors during
+  the last housekeeping sweep (every 2 seconds)".
+    * TODO: do we want to have a label to distinguish each error reason,
+      e.g. "Pod found, but volumes are still mounted on disk" from say
+      "orphaned pod %q found, but error occurred during reading of
+      volume-subpaths dir from disk"?
```

Comment:
to expand on the TODO item with an example. `cleanupOrphanedPodDirs` can fail with:

* `Orphaned pod found, but volumes are not cleaned up`
* `orphaned pod %q found, but error occurred during reading the pod dir from disk`
* `Orphaned pod found, but failed to remove volumes subdir`
* `orphaned pod %q found, but error occurred when trying to remove subdir %q: %v"`
* `orphaned pod %q found, but error occurred when trying to remove the pod directory: %v`
* `orphaned pod %q found, but failed to rmdir() volume at path %v: %v`
* `orphaned pod %q found, but error occurred during reading of volume-subpaths dir from disk: %v`
* `orphaned pod %q found, but error occurred when trying to remove the volumes dir: %v`

And potentially many more. Do we want to have a label `reason` with some enumerated values that would allow users to distinguish each error from each other or is it useless?